### PR TITLE
[7.0.x] Add controller service reconciler

### PR DESCRIPTION
**Pull request by bernardjkim**
## Description

This PR adds a controller service reconciler that will periodically attempt to reconcile the controller service. The cluster configuration resource now has an additional `gravityControllerService` spec that can be used to update the controller service config.

## Type of change

* Internal change (not necessarily a bug fix or a new feature)

## Linked tickets and other PRs

* Ports https://github.com/gravitational/gravity/pull/1995
* Requires https://github.com/gravitational/gravity/pull/2085

## TODOs

- [x] Self-review the change
- [x] Perform manual testing
- [ ] Address review feedback

## Testing done

Default gravity-site service config:
```
$ kubectl describe svc gravity-site -n kube-system
Name:                     gravity-site
Namespace:                kube-system
Labels:                   app=gravity-site
Annotations:              service.beta.kubernetes.io/aws-load-balancer-connection-idle-timeout: 3600
                          service.beta.kubernetes.op/aws-load-balancer-internal: 0.0.0.0/0
Selector:                 app=gravity-site
Type:                     LoadBalancer
IP:                       100.100.118.95
Port:                     web  3009/TCP
TargetPort:               3009/TCP
NodePort:                 web  32009/TCP
Endpoints:                167.172.195.128:3009
Session Affinity:         None
External Traffic Policy:  Cluster
Events:                   <none>
```
cluster-config.yaml
```
kind: ClusterConfiguration
version: v1
spec:
  global:
    featureGates:
      AllAlpha: false
  gravityControllerService:
    annotations:
      "cloud.google.com/load-balancer-type": "Internal"
    spec:
      type: NodePort
```

**Verify gravity-site service configuration is updated when updating cluster configuration**
- Exec `gravity resource create cluster-config.yaml`
- Verify gravity-site service config has been updated
```
$ kubectl describe svc gravity-site -n kube-system
Name:                     gravity-site
Namespace:                kube-system
Labels:                   app=gravity-site
Annotations:              cloud.google.com/load-balancer-type: Internal
Selector:                 app=gravity-site
Type:                     NodePort
IP:                       100.100.118.95
Port:                     web  3009/TCP
TargetPort:               3009/TCP
NodePort:                 web  32009/TCP
Endpoints:                167.172.195.128:3009
Session Affinity:         None
External Traffic Policy:  Cluster
Events:
  Type    Reason  Age   From                Message
  ----    ------  ----  ----                -------
  Normal  Type    78s   service-controller  LoadBalancer -> NodePort
```
**Verify gravity-site service configuration is updated on install**
- Exec `gravity install --config cluster-config.yaml`
- Verify gravity-site service is created with specified config
```
$ kubectl describe svc gravity-site -n kube-system
Name:                     gravity-site
Namespace:                kube-system
Labels:                   app=gravity-site
Annotations:              cloud.google.com/load-balancer-type: Internal
Selector:                 app=gravity-site
Type:                     NodePort
IP:                       100.100.118.199
Port:                     web  3009/TCP
TargetPort:               3009/TCP
NodePort:                 web  32009/TCP
Endpoints:                167.172.195.128:3009
Session Affinity:         None
External Traffic Policy:  Cluster
Events:
  Type    Reason  Age   From                Message
  ----    ------  ----  ----                -------
  Normal  Type    10m   service-controller  LoadBalancer -> NodePort
```
**Verify gravity-site service configuration is updated on upgrade**
- Upgrade gravity version
- Verify gravity-site service config changes persist after upgrade.
```
$ kubectl describe svc gravity-site -n kube-system
Name:                     gravity-site
Namespace:                kube-system
Labels:                   app=gravity-site
Annotations:              cloud.google.com/load-balancer-type: Internal
Selector:                 app=gravity-site
Type:                     NodePort
IP:                       100.100.118.199
Port:                     web  3009/TCP
TargetPort:               3009/TCP
NodePort:                 web  32009/TCP
Endpoints:                167.172.195.128:3009
Session Affinity:         None
External Traffic Policy:  Cluster
Events:
  Type    Reason  Age                  From                Message
  ----    ------  ----                 ----                -------
  Normal  Type    2m55s                service-controller  NodePort -> LoadBalancer
  Normal  Type    2m21s (x2 over 56m)  service-controller  LoadBalancer -> NodePort
```

### CODE REVIEW

File: lib/storage/clusterconfig/kubernetes_test.go [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+}
+
+// Update updates the controller service.
+func (r *mockServiceControl) Update(config *GravityControllerService) error {
+	if r.svc == nil {
+		r.svc = newService(config)
+		return nil
+	}
+
+	if !shouldUpdate(toServiceConfig(r.svc), config) {
```
Comment (a-palchikov) [inline lib/storage/clusterconfig/kubernetes_test.go:322]: I don't think additional test functionality (esp. in context of mocks) is a helpful test - this is not used anywhere but the test and replaces the actual implementation. We should consider implementing this as a test to execute in the context of an actual cluster (eventually, we will be running tests that require a cluster and we already have a bunch of them in the repository) using the actual implementation of the service control and cluster config control implementations.

File: lib/storage/clusterconfig/utils.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+	if existing.Spec.Type != incoming.Spec.Type {
+		return true
+	}
+
+	if len(existing.Spec.Ports) != len(incoming.Spec.Ports) {
+		return true
+	}
+	for i, incomingPort := range incoming.Spec.Ports {
+		existingPort := existing.Spec.Ports[i]
```
Comment (a-palchikov) [inline lib/storage/clusterconfig/utils.go:85]: theoretically, the slices might be the same but out of order.

File: lib/storage/clusterconfig/clusterconfig.go [hunk 3]
```diff
@@ -140,6 +148,36 @@ func (r Resource) Merge(other Resource) Resource {
...
+				r.Spec.GravityControllerService.Annotations[k] = v
+			}
+		}
+
+		if updateGravityService.Spec.Type != "" {
+			r.Spec.GravityControllerService.Spec.Type = updateGravityService.Spec.Type
+		}
+
+		if len(updateGravityService.Spec.Ports) != 0 {
+			r.Spec.GravityControllerService.Spec.Ports = updateGravityService.Spec.Ports
```
[the remark this answers is not in the document]
Comment (bernardjkim) [reply lib/storage/clusterconfig/clusterconfig.go:177]: Looks like the returned Resource is immediately marshaled to JSON and then discarded. Copying via assignment should be okay here. Will update labels/annotations.

**Comment by r0mant** [comment]
Closing as well as https://github.com/gravitational/gravity/pull/1995.

**Pull request closed without merging**